### PR TITLE
CICD

### PR DESCRIPTION
**Pull request by michaeljfazio**
* Added GitHub workflow for CICD on push and merge requests (includes compilation check, linting, code formatting, and clippy checks).
* Update README.md to include contribution instructions and also specify recurse submodules on clone.
* Reformatted code according to standard rustfmt rules (enforced in CICD builds).
* Moved libsodium submodule to contrib directory and re-wrote build.rs to use autotools crate to simplify build code. 
* Added conditional compilation check for unit test failing on OSX due to differing expected error message.
* Update all out-of-date dependencies (including those with security notices). 
* Implemented workflow builds for both Ubuntu and macOS.

### CODE REVIEW

**Comment by michaeljfazio** [comment]
I'm not going to complete the release workflow as part of this pull request. That can be finished at a later stage. Work remaining is to upload the matrix build artefacts (i.e. cncli binaries) automatically.

File: README.md [hunk 1]
```diff
@@ -18,20 +18,29 @@ $ source $HOME/.cargo/env
...
-$ rustup update  
+$ rustup update
+$ rustup component add clippy rustfmt
 ```
 #### Install dependencies and build cncli
 ```shell script
 $ source $HOME/.cargo/env
 $ sudo apt-get update -y
-$ sudo apt-get install automake build-essential pkg-config libffi-dev libgmp-dev libssl-dev libtinfo-dev libsystemd-dev zlib1g-dev make g++ tmux git jq wget libncursesw5 libtool autoconf -y
+$ sudo apt-get install -y automake build-essential pkg-config libffi-dev libgmp-dev libssl-dev libtinfo-dev libsystemd-dev zlib1g-dev make g++ tmux git jq wget libncursesw5 libtool autoconf libsodium-dev
```
Comment (AndrewWestberg) [inline README.md:28]: I don't think this is correct. We must use the iohk version of libsodium and not the system-installed version. The system-installed version is missing functions we need.

File: build.rs [hunk 2]
```diff
@@ -1,49 +1,11 @@
...
-            .arg("--hard")
-            .arg("HEAD")
-            .current_dir("libsodium")
-    });
-
-    run("./autogen.sh", |command| { command.current_dir("libsodium") });
-    run("./configure", |command| { command.current_dir("libsodium") });
-    run("make", |command| { command.current_dir("libsodium") });
-    println!(r"cargo:rustc-link-search=libsodium/src/libsodium/.libs");
+    pkg_config::Config::new().probe("libsodium").unwrap();
```
Comment (AndrewWestberg) [inline build.rs:9]: If this is relying on the system-installed version of libsodium, it's not the right approach.

**Review by AndrewWestberg — CHANGES_REQUESTED** [review]
CICD stuff looks good. We need to figure out the right approach for libsodium before merging.

**Comment by michaeljfazio** [comment]
> CICD stuff looks good. We need to figure out the right approach for libsodium before merging.

All done. I'll update the pull request details accordingly.

**Review by AndrewWestberg — APPROVED** [review]

**Pull request merged**